### ISSUE TITLE
"Transport header does not contain interleaved IDs" with specific TCP client

### ISSUE THREAD

**Issue by neilyoung** [aler9/mediamtx#1650]
## Which version are you using?

v0.21.1

## Which operating system are you using?

- [ ] Linux amd64 standard
- [ ] Linux amd64 Docker
- [x] Linux arm64 standard
- [ ] Linux arm64 Docker
- [ ] Linux arm7 standard
- [ ] Linux arm7 Docker
- [ ] Linux arm6 standard
- [ ] Linux arm6 Docker
- [ ] Windows amd64 standard
- [ ] Windows amd64 Docker (WSL backend)
- [ ] macOS amd64 standard
- [ ] macOS amd64 Docker
- [ ] Other (please describe)

## Describe the issue

Customer (Australia) uses "Genetec Video System to receive the RTSP". They are unable to receive video via TCP using this system. UDP is reported to work. VLC and FFPLAY is reported to work too. With the Genetec system sometimes the DESCRIBE request is answered with "BAD REQUEST", sometimes "OK". Even in OK case there is no video (flowing).

The only indication that something is wrong: I see a lot of traces of this kind in the RTSP server log:

```
2023/04/05 07:21:23 INF [RTSP] [session 028d421f] created by 202.xxx.xxx.xxx:34140
2023/04/05 07:21:23 INF [RTSP] [conn 202.xxx.xxx.xxx:34140] closed (transport header does not contain interleaved IDs)
2023/04/05 07:21:23 INF [RTSP] [session 028d421f] destroyed (not in use)
```

## Describe how to replicate the issue

Replication difficult, I also don't have that "video system".  I just have the Wireshark traces taken by them.

## Did you attach the server logs?

No

## Did you attach a network dump?

I have that, but I would not like to publish it here. Can provide it using a private dropbox link, if I would have a way to securely communicate that.

Sorry for the "no-information". Hoping we could get in touch with this.

**Comment by neilyoung** [comment]
I will be getting access to such a system later today. Would you consider dealing with this issue then?

**Comment by aler9** [comment]
Hello, send server logs with `logLevel: debug` and network dump to aler9.dev@gmail.com

**Comment by neilyoung** [comment]
Thanks. Will do when I get some fresh ones.

**Comment by neilyoung** [comment]
Hi Alessandro,

Thanks for your offer to have a look at it. 

OK, situation recap: We have a Sydney based client, who is fixed on using some VMS they have to use. Its name is Genetec Video System. It identifies as "OmnicastRTSPClient 1.0" and has several issues with Simple RTSP server (or vice versa).

In order to not bother the client with things probably caused by us we asked the Genetec people to provide us a test version and finally we managed to install it on a Windows machine (in Miami). Besides the already reported BAD REQUEST response from the RTSP server we are using (your version 0.20), when it is asking to provide a TCP stream, I actually see yet another problem with UDP, which manifests in an immediate close of the RTSP session between client and server from client side or in a delayed TEARDOWN from the client after a couple of seconds. I think it is, because they simply cannot come together: The client demands 15000-15001 as port, the server gives 8000-8001. No way to meet, I guess (everything under reserve that I have understood correctly)... 

DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
CSeq: 1
User-Agent: OmnicastRTSPClient/1.0
Accept: application/sdp

RTSP/1.0 200 OK
CSeq: 1
Content-Base: rtsp://3.83.200.4:8554/1567467946/test/
Content-Length: 279
Content-Type: application/sdp
Server: gortsplib

v=0
o=- 0 0 IN IP4 127.0.0.1
s=Stream
c=IN IP4 0.0.0.0
t=0 0
m=video 0 RTP/AVP 96
a=control:mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1; sprop-parameter-sets=Z0LAIIyNQCgC3QDwiEag,aM48gA==; profile-level-id=42C020
SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
CSeq: 2
User-Agent: OmnicastRTSPClient/1.0
Transport: RTP/AVP;unicast;client_port=15000-15001

RTSP/1.0 200 OK
CSeq: 2
Server: gortsplib
Session: deff3b38-c855-4666-922b-4ccb2a9a3d91;timeout=60
Transport: RTP/AVP;unicast;client_port=15000-15001;server_port=8000-8001;ssrc=E8756B92

PLAY rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
CSeq: 3
Session: deff3b38-c855-4666-922b-4ccb2a9a3d91
User-Agent: OmnicastRTSPClient/1.0
Range: npt=0.000-

RTSP/1.0 200 OK
CSeq: 3
RTP-Info: url=rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783;seq=3935;rtptime=1393610235
Server: gortsplib
Session: deff3b38-c855-4666-922b-4ccb2a9a3d91;timeout=60

TEARDOWN rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
CSeq: 4
Session: deff3b38-c855-4666-922b-4ccb2a9a3d91
User-Agent: OmnicastRTSPClient/1.0

RTSP/1.0 200 OK
CSeq: 4
Server: gortsplib



Besides the fact, that we could generally workaround by opening these two 15k ports, I have a general question: How to proceed here? I don't recall to have seen that these ports have to be opened and seemingly it is up to the client to choose (?). However, now every XY client can come an ask for another pair? How could I adapt with my firewall?

Anyway, the time this test was executed the firewall was down, so a possible failing communication via port 15000/15001 cannot be caused by firewall issues. 

The server log (sorry, just have it as plain text here) for an Unicast UDP connection attempt (no video flows) is here:

023/04/11 05:49:59 INF [RTSP] [conn 75.46.62.166:50771] opened
2023/04/11 05:49:59 DEB [RTSP] [conn 75.46.62.166:50771] [c->s] DESCRIBE rtsp://ms.votix.com:8554/1567467946/test RTSP/1.0
2023/04/11 05:49:59 DEB [RTSP] [conn 75.46.62.166:50771] [s->c] RTSP/1.0 200 OK
2023/04/11 05:49:59 INF [RTSP] [conn 75.46.62.166:50771] closed (EOF)
2023/04/11 05:50:05 INF [RTSP] [conn 75.46.62.166:50781] opened
2023/04/11 05:50:05 DEB [RTSP] [conn 75.46.62.166:50781] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:05 DEB [RTSP] [conn 75.46.62.166:50781] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:05 INF [RTSP] [conn 75.46.62.166:50781] closed (EOF)
2023/04/11 05:50:13 INF [RTSP] [conn 75.46.62.166:50796] opened
2023/04/11 05:50:13 DEB [RTSP] [conn 75.46.62.166:50796] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:13 DEB [RTSP] [conn 75.46.62.166:50796] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:13 DEB [RTSP] [conn 75.46.62.166:50796] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:50:13 INF [RTSP] [session 2f3da72c] created by 75.46.62.166:50796
2023/04/11 05:50:13 DEB [RTSP] [conn 75.46.62.166:50796] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:14 DEB [RTSP] [conn 75.46.62.166:50796] [c->s] PLAY rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:14 DEB [RTSP] [conn 75.46.62.166:50796] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:23 DEB [RTSP] [conn 75.46.62.166:50796] [c->s] TEARDOWN rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:23 INF [RTSP] [session 2f3da72c] destroyed (teared down by 75.46.62.166:50796)
2023/04/11 05:50:23 DEB [RTSP] [conn 75.46.62.166:50796] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:23 INF [RTSP] [conn 75.46.62.166:50796] closed (EOF)
2023/04/11 05:50:23 INF [RTSP] [conn 75.46.62.166:50817] opened
2023/04/11 05:50:23 DEB [RTSP] [conn 75.46.62.166:50817] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:23 DEB [RTSP] [conn 75.46.62.166:50817] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:23 DEB [RTSP] [conn 75.46.62.166:50817] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:50:23 INF [RTSP] [session b74dd555] created by 75.46.62.166:50817
2023/04/11 05:50:23 DEB [RTSP] [conn 75.46.62.166:50817] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:24 DEB [RTSP] [conn 75.46.62.166:50817] [c->s] PLAY rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:24 DEB [RTSP] [conn 75.46.62.166:50817] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:27 DEB [RTSP] [conn 75.46.62.166:50817] [c->s] TEARDOWN rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:50:27 INF [RTSP] [session b74dd555] destroyed (teared down by 75.46.62.166:50817)
2023/04/11 05:50:27 DEB [RTSP] [conn 75.46.62.166:50817] [s->c] RTSP/1.0 200 OK
2023/04/11 05:50:27 INF [RTSP] [conn 75.46.62.166:50817] closed (EOF)
The accompanying Wireshark log taken at client side is this:

https://www.dropbox.com/s/kdpu2namlql3mf5/unicastudp.zip?dl=1


Now to the BAD REQUEST thing on TCP requests.

Here the server log, which pretty much directly says, what's wrong (transport header does not contain interleaved IDs), even though I wouldn't know what's going on:


2023/04/11 05:52:13 INF [RTSP] [conn 75.46.62.166:51005] opened
2023/04/11 05:52:13 DEB [RTSP] [conn 75.46.62.166:51005] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:13 DEB [RTSP] [conn 75.46.62.166:51005] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:16 INF [RTSP] [conn 75.46.62.166:51005] closed (EOF)
2023/04/11 05:52:16 INF [RTSP] [conn 75.46.62.166:51011] opened
2023/04/11 05:52:16 DEB [RTSP] [conn 75.46.62.166:51011] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:16 DEB [RTSP] [conn 75.46.62.166:51011] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51011] closed (EOF)
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51029] opened
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51029] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51029] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51029] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:26 INF [RTSP] [session f2d393dc] created by 75.46.62.166:51029
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51029] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51029] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51030] opened
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51030] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51030] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51030] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:26 INF [RTSP] [session f07e25e8] created by 75.46.62.166:51030
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51030] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51030] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51031] opened
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51031] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51031] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51031] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:26 INF [RTSP] [session f6fc6d11] created by 75.46.62.166:51031
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51031] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51031] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51032] opened
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51032] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51032] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51032] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:26 INF [RTSP] [session a67a6189] created by 75.46.62.166:51032
2023/04/11 05:52:26 DEB [RTSP] [conn 75.46.62.166:51032] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:26 INF [RTSP] [conn 75.46.62.166:51032] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51033] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51033] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51033] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51033] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session 2cdfa003] created by 75.46.62.166:51033
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51033] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51033] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51034] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51034] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51034] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51034] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session e2f26dc2] created by 75.46.62.166:51034
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51034] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51034] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51036] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51036] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51036] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51036] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session b18a5377] created by 75.46.62.166:51036
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51036] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51036] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51037] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51037] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51037] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51037] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session c53e3386] created by 75.46.62.166:51037
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51037] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51037] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51038] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51038] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51038] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51038] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session eea002b2] created by 75.46.62.166:51038
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51038] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51038] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51039] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51039] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51039] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51039] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session a13828b4] created by 75.46.62.166:51039
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51039] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51039] closed (transport header does not contain interleaved IDs)
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51040] opened
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51040] [c->s] DESCRIBE rtsp://3.83.200.4:8554/1567467946/test RTSP/1.0
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51040] [s->c] RTSP/1.0 200 OK
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51040] [c->s] SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
2023/04/11 05:52:27 INF [RTSP] [session d19e6bbb] created by 75.46.62.166:51040
2023/04/11 05:52:27 DEB [RTSP] [conn 75.46.62.166:51040] [s->c] RTSP/1.0 400 Bad Request
2023/04/11 05:52:27 INF [RTSP] [conn 75.46.62.166:51040] closed (transport header does not contain interleaved IDs)
And the wireshark log:

https://www.dropbox.com/s/u8pr8pop9s5thvq/rtspovertcp.zip?dl=1


Here the question is: What makes you rejecting the SETUP?


SETUP rtsp://3.83.200.4:8554/1567467946/test/mediaUUID=bab0b78b-5042-4e7a-bb57-7ea981df2783 RTSP/1.0
CSeq: 2
User-Agent: OmnicastRTSPClient/1.0
Transport: RTP/AVP/TCP;unicast

RTSP/1.0 400 Bad Request
CSeq: 2
Server: gortsplib


Kind regards

**Comment by neilyoung** [comment]
Sent it

**Comment by neilyoung** [comment]
So this error message "Transport header does not contain interleaved IDs" does have its origin in gosrtplib (https://github.com/bluenviron/gortsplib). 

gosrtplib obviously unconditionally expects the `Transport` header of a TCP setup to look like so or similar


https://github.com/bluenviron/gortsplib/blob/main/client.go#L1410

```
SETUP rtsp://foo.com/bar.file RTSP/1.0
           CSeq: 2
           Transport: RTP/AVP/TCP;interleaved=0-1
```

This is not the case from the Genetec. According to the RFC https://www.rfc-editor.org/rfc/rfc2326.html#section-12.39 the parameter is mandatory for TCP, not UDP. It separates the various streams (0 for Audio and 1 for Video in the above case).

**Comment by neilyoung** [comment]
This is a SETUP send by FFPLAY:

```
SETUP rtsp://ms.votix.com:8554/6227329315/vx/mediaUUID=fafe08b6-047b-4e25-a9c5-6bc1208ba0ac RTSP/1.0
Transport: RTP/AVP/TCP;unicast;interleaved=0-1
CSeq: 4
User-Agent: Lavf59.27.100
Authorization: Basic dm90aXg6dm90aXg=

```

You see it?

**Comment by aler9** [comment]
Hello,

Regarding UDP, handshake between the NVR and the server is working as expected, but, communication can't be established because UDP only works when server and client are on the same LAN.

Regarding TCP, please test this nightly release and let me know if it fixes the issue. If it doesn't, provide server logs with `logLevel: debug` option.

[mediamtx_v0.22.0-16-gf3a728b_darwin_amd64.tar.gz](https://github.com/aler9/mediamtx/files/11220945/mediamtx_v0.22.0-16-gf3a728b_darwin_amd64.tar.gz)
[mediamtx_v0.22.0-16-gf3a728b_linux_amd64.tar.gz](https://github.com/aler9/mediamtx/files/11220949/mediamtx_v0.22.0-16-gf3a728b_linux_amd64.tar.gz)
[mediamtx_v0.22.0-16-gf3a728b_linux_arm64v8.tar.gz](https://github.com/aler9/mediamtx/files/11220953/mediamtx_v0.22.0-16-gf3a728b_linux_arm64v8.tar.gz)
[mediamtx_v0.22.0-16-gf3a728b_linux_armv6.tar.gz](https://github.com/aler9/mediamtx/files/11220957/mediamtx_v0.22.0-16-gf3a728b_linux_armv6.tar.gz)
[mediamtx_v0.22.0-16-gf3a728b_linux_armv7.tar.gz](https://github.com/aler9/mediamtx/files/11220961/mediamtx_v0.22.0-16-gf3a728b_linux_armv7.tar.gz)
[mediamtx_v0.22.0-16-gf3a728b_windows_amd64.zip](https://github.com/aler9/mediamtx/files/11220963/mediamtx_v0.22.0-16-gf3a728b_windows_amd64.zip)

**Comment by neilyoung** [comment]
> Regarding UDP, handshake between the NVR and the server is working as expected, but communication can't be established because UDP only works when server and client are on the same LAN.

Are you sure with this `UDP only works when server and client are on the same LAN.`? Could you kindly elaborate? I'm asking because up to now I was under the impression, that this SETUP is **really** setting up an UDP communication for the RTP (a bit anonymized):

```
SETUP rtsp://******:8554/6227329315/vx/mediaUUID=b719d37a-a808-436a-b916-5d5eb41a0552 RTSP/1.0
Transport: RTP/AVP/UDP;unicast;client_port=34836-34837
CSeq: 4
User-Agent: Lavf59.27.100
Authorization: Basic ******

RTSP/1.0 200 OK
CSeq: 4
Server: gortsplib
Session: a519ce47-9044-42f2-8404-e314613cf392;timeout=60
Transport: RTP/AVP;unicast;client_port=34836-34837;server_port=8000-8001;ssrc=D42CC60C
```

And in fact - **it works**, even though the server is somewhere at AWS and my client is here.

The FFPLAY command line is

```
ffplay  rtsp://xxx:xxx@*******:8554/6227329315/vx
```

In addition, I'm perfectly able to establish a TCP RTP session with this SETUP sequence:

```
SETUP rtsp://******:8554/6227329315/vx/mediaUUID=b719d37a-a808-436a-b916-5d5eb41a0552 RTSP/1.0
Transport: RTP/AVP/TCP;unicast;interleaved=0-1
CSeq: 4
User-Agent: Lavf59.27.100
Authorization: Basic ******

RTSP/1.0 200 OK
CSeq: 4
Server: gortsplib
Session: c16d3bf5-5e12-4b61-9ab1-0bc6538765de
Transport: RTP/AVP/TCP;unicast;interleaved=0-1;ssrc=D42CC60C

```

The matching FFPLAY command line is this:

```
ffplay  rtsp://xxx:xxx@*******:8554/6227329315/vx -rtsp_transport tcp
```

Thanks for the code provided. I will let you know. I saw that you made a PR to the gorrtsplib, so I'm sure it will work. But isn't this `interleaved` info element for a TCP session mandatory? I checked the RFC and couldn't find a hint that this would be optional.

**Comment by aler9** [comment]
> UDP only works when server and client are on the same LAN.? 

UDP also works when the server has a public IP, like an AWS elastic IP. The issue is not the handshake, that happens by using a TCP session, but the streaming, which happens through UDP packets. If there's a NAT/firewall in between, there's a high probability that these UDP packets gets their source port changed or get blocked.

Anyway, the topic of this issue is the ability to support TCP readers that don't provide interleaved IDs in requests. For any other question, you can open additional issues or discussions.

Regarding standards and specifications, there are a lot of RTSP devices that don't follow rules and can't be changed since they are embedded hardware. Supporting or not supporting them depends on evaluating whether the change required for them to work impacts other devices or has security implications.

Just test the nightly release and report whether the fix is enough to solve the issue.

**Comment by neilyoung** [comment]
> Anyway, the topic of this issue is the ability to support TCP readers that don't provide interleaved IDs in requests. For any other question, you can open additional issues or discussions.

Sure. I was just stumbling over that UDP only in the same LAN claim. But I'm also sure, that AWS has NAT applied. Even though it works. I think the port address translation applied to NAT does only apply, if you want to stream UDP towards the RTSP server through NAT. Here in this case it is the other way around and if the client demands a specific port and the server follows: What can go wrong?

In another UDP case the client did setup 

<img width="497" alt="image" src="https://user-images.githubusercontent.com/731020/231754926-9973469b-d2f0-4dd1-bb55-fca2c306f65e.png">

And this was the result from AWS to me:

<img width="554" alt="image" src="https://user-images.githubusercontent.com/731020/231754752-82954996-4f78-4733-a7a4-20a14d9a16de.png">

I don't see the problem here with NAT. But agreed, this is another party.

**Comment by aler9** [comment]
added in MediaMTX v0.22.1

**Comment by neilyoung** [comment]
BTW: FYI. The Genetec UDP issue is a missing UDP hole punching, which is done by VLC and Co. Will most likely not work in all cases, but in many. But here 2 times 4 bytes are sufficient to make the way free for the RTP from the server downstream.

**Comment by aler9** [comment]
@neilyoung the server performs UDP  hole punching too, in the same exact way

**Comment by neilyoung** [comment]
I will forward this info to Genetec. Hoping they consider to adopt.

**Comment by neilyoung** [comment]
> @neilyoung the server performs UDP hole punching too, in the same exact way

Out of curiosity: Could you provide me a code link, where exactly the server is doing this? It could help me in my discussion with Genetec

**Comment by aler9** [comment]
A small clarification: when i said that "the server performs UDP hole punching" i meant that MediaMTX performs UDP hole punching when reading a stream from a RTSP camera or from a RTSP client. It's the reader that must implement UDP hole punching.

In your case it's the Genetec NVR that must implement UDP hole punching.

here's the code:

https://github.com/bluenviron/gortsplib/blob/5258f5b6e69254edb0dcb4958b0bf24ce8af5dd3/client.go#L1497-L1509

**Comment by neilyoung** [comment]
Yes. Thanks. I’m aware of the roles. Thanks for the link. I already digested my results and will forward it to Genetec in the hope that they’ll fix it. I also already tested several other clients, namely FFPLAY, GStreamer and VLC and they are all doing this UDP hole punching in one way or another. And also the mandatory interlaced info element for TCP/TLS is sent by them. So the ball is in the field of Genetc. I hope they’ll see that too. 

At least you have saved our solution by enabling TCP with them for our clients. This wass a boss move. 😀. Thanks again. I love your project.
